### PR TITLE
Issue29 configurable via config yaml file

### PR DESCRIPTION
**Pull request by hewison-chris**
This adds a configuration section for Faucet as a web service and as a transaction generator.

### CODE REVIEW

File: doc/config.example.yaml [hunk 1]
```diff
@@ -1,35 +1,61 @@
...
-  - SBTL7LWIEMOVCPFFEKJLGYM6ZVE5GXCUX3ACI262YMVFWHJ2SQDABI4B
-  - SA63FZ3KRHA44PY56PTF44MX55MMJJHRR7TV4REEYAH6IGR5NSIA3FN4
-  - SAC6GZSDGDXF447OJK5MU6SAKVOP5WONYAIV2TIOFGXX2GPKUXZQ4ZAJ
-  - SAZWRKDD43DC5ETOW4L66VZK2U6RVAXHM7XJLAASYCMBQVQQ335AG7JA
+stats_port: 9113
+
+################################################################################
+##                  Transaction Generator Configuration                       ##
+################################################################################
+generator:
```
Comment (linked0) [inline doc/config.example.yaml:12]: Isn't it better to be `tx_generator`?

File: source/faucet/main.d [hunk 2]
```diff
@@ -56,26 +56,76 @@ import vibe.web.rest;
 
 static immutable KeyCount = WK.Keys.byRange().length;
 
-/// Configuration parameter for Faucet
-private struct Config
+private struct Generator
```
Comment (linked0) [inline source/faucet/main.d:59]: How about `TxGenerator`?

File: source/faucet/main.d [hunk 3]
```diff
@@ -36,7 +36,7 @@ import dyaml.loader;
 
 import std.algorithm;
 import std.exception;
-static import std.file;
+import std.file;
```
Comment (kchulj) [inline source/faucet/main.d:39]: I guess we don't need `static` ?

File: source/faucet/main.d [hunk 4]
```diff
@@ -629,15 +695,21 @@ private const(string)[] parseSequence (string section,
...
+        Generator(root["generator"]),
+        root["web_bind_address"].as!string,
+        root["web_bind_port"].as!ushort,
+        root["stats_port"].as!ushort);
 }
 
 /// Global because we need to access it from our signal handler
 private Faucet inst;
 
+private SecretKey[PublicKey] secret_keys;
```
Comment (linked0) [inline source/faucet/main.d:711]: This should be in the upper part of the file because the variable is being used in the above part of the file.

File: source/faucet/main.d [hunk 5]
```diff
@@ -266,10 +316,12 @@ public class Faucet : FaucetAPI
 
     ***************************************************************************/
 
-    public this (const Config config, const Address address)
+     public this (in Config config)
```
Comment (linked0) [inline source/faucet/main.d:319]: Unnecessary space.

**Review by linked0 — APPROVED** [review]
Great, some nits.

**Comment by linked0** [comment]
Never mind for requesting changing the name 'Generator` and 'generator' if you are not doing something for that. `Faucet` is already for generating `tx`s, I think.

**Comment by kchulj** [comment]
> Never mind for requesting changing the names, `Generator` and `generator` if you are not doing something for that.
> `Faucet` is already for generating `tx`s, I think.

The generator is generating `tx`s so I think it was a fair suggestion.
AFAIK the faucet only generates `tx`s when someone requests coins.

File: source/faucet/main.d (hunk 3 above)
Comment (hewison-chris) [reply source/faucet/main.d:39]: Not sure why it was `static import` to start with but I wanted to use `.exists` after the name of the config file so it would not work as it was.

File: source/faucet/main.d (hunk 4 above)
Comment (kchulj) [reply source/faucet/main.d:711]: Oh why did this become global?

File: doc/config.example.yaml [hunk 6]
```diff
@@ -1,35 +1,61 @@
...
+  # How frequently we run our periodic task in seconds
+  send_every_seconds: 1
```
Comment (Geod24) [inline doc/config.example.yaml:14]: Frequency, interval, etc... would be better names

File: doc/config.example.yaml [hunk 7]
```diff
@@ -1,35 +1,61 @@
...
+
+  # Between how many addresses we split a transaction by
+  split_count: 1
+
+  # Maximum number of utxo before merging instead of splitting
+  max_utxos: 10
```
Comment (Geod24) [inline doc/config.example.yaml:20]: merge_threshold ?

File: doc/config.example.yaml [hunk 8]
```diff
@@ -1,35 +1,61 @@
...
-  - SCCQRMDR63E6I5QPCYMQAXQ2NYULG562PKPLVSNTKQ6CEDGPXXWAYZQE
-  - SARGARQL5O7K7AGVPJ6W2MLUEKAKFCPCSP5P7O7ZV25GGBQE5VKAN2CT
-  - SDHNZA6CWQHHC2ZIZBAC4KZIJG2KL45RMGVPWQ226C6HQYX6N6ZQ7KDI
-  - SDAFJKB4IDJ5FRP5L2PL2TI6RZFVXBCNG36YW2YE3OM2UJTWOANQOID3
-  - SDD6TBCC7FTPJWMX5NJ5F2SVS5T3GA74KN7YWMEX3ALROS7PJ72QOOCM
-  - SBTL7LWIEMOVCPFFEKJLGYM6ZVE5GXCUX3ACI262YMVFWHJ2SQDABI4B
-  - SA63FZ3KRHA44PY56PTF44MX55MMJJHRR7TV4REEYAH6IGR5NSIA3FN4
-  - SAC6GZSDGDXF447OJK5MU6SAKVOP5WONYAIV2TIOFGXX2GPKUXZQ4ZAJ
-  - SAZWRKDD43DC5ETOW4L66VZK2U6RVAXHM7XJLAASYCMBQVQQ335AG7JA
+stats_port: 9113
```
Comment (Geod24) [inline doc/config.example.yaml:7]: No global level variable please.

File: source/faucet/main.d (hunk 3 above)
Comment (Geod24) [reply source/faucet/main.d:39]: I usually `static import std.file` because the names it provides are very generic. But it's a personal habit, so I don't mind changing it.

File: source/faucet/main.d (hunk 5 above)
Comment (Geod24) [reply source/faucet/main.d:319]: Also it's being stored in the class so by definition it is not `scope`.

File: source/faucet/main.d [hunk 9]
```diff
@@ -347,7 +399,7 @@ public class Faucet : FaucetAPI
 
     public void setup (uint count)
     {
-        while (!this.state.update(this.client, Height(0)))
+        while (!this.state.update(this.clients[0], Height(0)))
```
Comment (Geod24) [inline source/faucet/main.d:402]: Would have been better to separate support for multiple clients in its own commit.

File: source/faucet/main.d [hunk 10]
```diff
@@ -524,24 +564,14 @@ int main (string[] args)
     {
         stderr.writeln("Usage: ./faucet <address>");
```
Comment (Geod24) [inline source/faucet/main.d:565]:
```suggestion
stderr.writeln("Usage: ./faucet -c {'none'|<path>} [<address>]");
```

File: source/faucet/main.d (hunk 4 above)
Comment (hewison-chris) [reply source/faucet/main.d:711]: I update them before creating the Faucet instance.

File: doc/config.example.yaml (hunk 6 above)
Comment (hewison-chris) [reply doc/config.example.yaml:14]: How about `send_interval`?

File: doc/config.example.yaml (hunk 8 above)
Comment (hewison-chris) [reply doc/config.example.yaml:7]: What about `stats port` ?

File: source/faucet/main.d (hunk 9 above)
Comment (hewison-chris) [reply source/faucet/main.d:402]: Split the multi-client support into separate commit.

File: source/faucet/main.d (hunk 10 above)
Comment (hewison-chris) [reply source/faucet/main.d:565]: Updated

File: doc/config.example.yaml (hunk 8 above)
Comment (hewison-chris) [reply doc/config.example.yaml:7]: I created `Web` section for address and port. Decided to put `stats_port` in `tx_generator` section.

**Comment by hewison-chris** [comment]
I made the config global as I don't see why we need to pass it into the Faucet class.

File: source/faucet/main.d [hunk 11]
```diff
@@ -444,7 +458,7 @@ public class Faucet : FaucetAPI
             this.setup(config.tx_generator.split_count);
 
         // For now we always send to first client
```
Comment (kchulj) [inline source/faucet/main.d:460]: I think this comment can be removed.

File: source/faucet/main.d (hunk 5 above)
Comment (hewison-chris) [reply source/faucet/main.d]: Made it global.

File: source/faucet/main.d [hunk 12]
```diff
@@ -56,26 +56,90 @@ import vibe.web.rest;
 
 static immutable KeyCount = WK.Keys.byRange().length;
 
-/// Configuration parameter for Faucet
-private struct Config
+/// The keys that will be used for generating transactions
+private SecretKey[PublicKey] secret_keys;
+
+/// The configuratkion for faucet as a faucet and a tx generator
```
Comment (kchulj) [inline source/faucet/main.d:62]:
```suggestion
/// The configuration for faucet as a faucet and a tx generator
```

File: source/faucet/main.d [hunk 13]
```diff
@@ -56,26 +56,90 @@ import vibe.web.rest;
...
+        merge_threshold = node["merge_threshold"].as!uint;
+        () @trusted { addresses = parseSequence("addresses", node, true); }();
+        () @trusted { keys = parseSequence("keys", node, false); }();
+    }
+}
+
+/// The Config for the faucet web
+private struct Web
+{
+/// Address to bind for website
```
Comment (kchulj) [inline source/faucet/main.d:110]:
```suggestion
    /// Address to bind for website
```

File: source/faucet/main.d [hunk 14]
```diff
@@ -56,26 +56,90 @@ import vibe.web.rest;
...
+    {
+        this.stats_port = stats_port;
+        this.send_interval = send_interval;
+        this.split_count = split_count;
+        this.merge_threshold = merge_threshold;
+        this.addresses = addresses;
+        this.keys = keys;
+    }
+
+    this (Node node) @safe
```
Comment (kchulj) [inline source/faucet/main.d:96]: is this a `yaml` Node?

Comment (hewison-chris) [reply source/faucet/main.d:96]: Yes, I'll update the variable name to make this clearer.

Comment (kchulj) [reply source/faucet/main.d]: Oh I just asked to make sure. The name change might not have been necessary because the name `node` is used in other places as well 🤔

Comment (hewison-chris) [reply source/faucet/main.d]: See the last commit :-)

Comment (hewison-chris) [reply source/faucet/main.d]: I think it was a good idea to rename as we normally have a different node in mind when we read `node`.

File: source/faucet/main.d [hunk 15]
```diff
@@ -56,26 +56,90 @@ import vibe.web.rest;
...
+/// The Config for the faucet web
+private struct Web
+{
+    /// Address to bind for website
+    string address;
+
+    /// Port to bind for website
+    ushort port;
+
+    this (string address, ushort port = 9113)
```
Comment (hewison-chris) [inline source/faucet/main.d:116]: This port default should not be 9113 as it clashes with stats default port. I'll update it.

Comment (hewison-chris) [reply source/faucet/main.d]: Updated

Comment (kchulj) [reply source/faucet/main.d]: Good catch..

File: doc/config.example.yaml [hunk 16]
```diff
@@ -1,35 +1,64 @@
...
+  stats_port: 9113
+
+  # How frequently we run our periodic task in seconds
+  send_interval: 1
+
+  # Between how many addresses we split a transaction by
+  split_count: 1
+
+  # Maximum number of utxo before merging instead of splitting
+  merge_threshold: 10
```
Comment (kchulj) [inline doc/config.example.yaml:23]: The threshold seems to be a bit low. Should we increase?

Comment (hewison-chris) [reply doc/config.example.yaml:23]: This is just an example file so IMO it does not really matter too much.

**Comment by hewison-chris** [comment]
The last commit splits the configuration from the `main.d` file.

File: source/faucet/main.d [hunk 17]
```diff
@@ -329,12 +329,26 @@ public class Faucet : FaucetAPI
...
 
+    /*******************************************************************************
+
+        Take one of the clients selecting it randomly
+
+        Returns:
+          A client to send transactions or requests
+
+    *******************************************************************************/
+    private API randomClient () @trusted
```
Comment (Geod24) [inline source/faucet/main.d:265]: Is `choice` not `@safe` ?

File: source/faucet/main.d [hunk 18]
```diff
@@ -329,12 +329,26 @@ public class Faucet : FaucetAPI
...
     }
 
+    /*******************************************************************************
+
+        Take one of the clients selecting it randomly
+
+        Returns:
+          A client to send transactions or requests
+
+    *******************************************************************************/
```
Comment (Geod24) [inline source/faucet/main.d:264]: Nit: Newline

File: source/faucet/main.d [hunk 19]
```diff
@@ -408,7 +422,7 @@ public class Faucet : FaucetAPI
 
     public void setup (uint count)
     {
-        while (!this.state.update(this.client, Height(0)))
+        while (!this.state.update(randomClient(), Height(0)))
```
Comment (Geod24) [inline source/faucet/main.d:343]: Nit: `this`

File: source/faucet/main.d (hunk 17 above)
Comment (hewison-chris) [reply source/faucet/main.d:265]: No unfortunately not.

**Review by Geod24 — APPROVED** [review]
Nice. Not too found of have 3 global variables though, the reasons we have globals are:
1) For signal handling (hence Faucet);
2) For `Builder` (the keys);

The rest really does not need to be global.

**Comment by hewison-chris** [comment]
> Nice. Not too found of have 3 global variables though, the reasons we have globals are:
> 
> 1. For signal handling (hence Faucet);
> 2. For `Builder` (the keys);
> 
> The rest really does not need to be global.

I can move them to Faucet but note that Builder will still access `secret_keys` via the global `inst` like it was before.

**Review by kchulj — APPROVED** [review]
LGTM.
It looks like `dub.json` needs an update?

**Comment by Geod24** [comment]
`git add source/faucet/config.d`

**Comment by hewison-chris** [comment]
Accidentally forgot to add the Config.d file to git and now have lost my local copy also. Not even reflog can help me this time.  Will have to recreate the file and push an update.

**Comment by hewison-chris** [comment]
Updated with config.d included and other requested changes except for removing the globals. It seems that we need access to them all over the file at the moment so will require a bit of restructure first.
